src/main/java/com/commercetools/sync/inventory/InventorySync.java

```diff
@@ -0,0 +1,380 @@
+     * supply channel reference that is replaced by reference pointing to {@code supplyChannelId}
+     * @param draft draft where reference should be replaced
+     * @param supplyChannelId ID of supply channel existing in target system
+     * @return {@link InventoryEntryDraft} with supply channel reference pointing to {@code supplyChannelId}
+     * and other data same as in {@code draft}
+     */
+    private InventoryEntryDraft ofEntryDraftPlusRefToSupplyChannel(@Nonnull final InventoryEntryDraft draft,
+                                                                   @Nonnull final String supplyChannelId) {
+        final Reference<Channel> supplyChannelRef = Channel.referenceOfId(supplyChannelId);
+        return InventoryEntryDraftBuilder.of(draft)
+                .supplyChannel(supplyChannelRef)
+                .build();
+    }
+
+    /**
+     * If there is no entry for given {@code key} in map of existing supply channels key-id,
+     * method tries to create supply channel of given {@code key}.
+     *
+     * @param key key of supply channel that seems to not exists in a system.
+     * @return {@link Optional} containing ID of created/found supply channel for given key,
+     * or empty {@link Optional} when operation failed.
+     */
+    private synchronized Optional<String> createMissingSupplyChannel(@Nonnull final String key) {
+        if (supplyChannelKeyToId.containsKey(key)) {
+            return Optional.of(supplyChannelKeyToId.get(key));
+        } else {
+            try {
+                final Channel newChannel = inventoryService.createSupplyChannel(key);
+                if (newChannel != null) {
+                    supplyChannelKeyToId.put(newChannel.getKey(), newChannel.getId());
+                    return Optional.of(newChannel.getId());
+                }
+            } catch (Exception ex) {
+                failSync(format("Failed to create new supply channel of key '%s'", key), ex);
+            }
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Given a reason message as {@link String} and {@link Throwable} exception, this method calls the optional error
+     * callback specified in the {@code syncOptions}. <br/>
+     * <strong>Note</strong> to the end of {@code message} following phrase will be joined:
+     * <pre> " in CTP project with key '_KEY_'"</pre>
+     * where _KEY_ is replaced by CTP project key, taken from CTP client configuration.
+     *
+     * @param message the reason of failure
+     * @param exception the exception that occurred, if any
+     */
+    private void failSync(@Nonnull final String message, @Nullable final Throwable exception) {
+        final String finalMessage = message + format(" in CTP project with key '%s.'",
+                this.syncOptions.getCtpClient().getClientConfig().getProjectKey());
+        syncOptions.applyErrorCallback(finalMessage, exception);
+    }
+}
```

Comment:
I updated the approach, and got rid of thread pool. 100d96ff515d13f645bc086a9c10fcb46881499d